discovery/swarmkit/swarmkit.go

```diff
@@ -0,0 +1,138 @@
+package swarmkit
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/discovery"
+	"github.com/docker/swarm/cluster"
+)
+
+const defaultAgentPort = "2375"
+
+// Discovery is exported.
+type Discovery struct {
+	heartbeat  time.Duration
+	ttl        time.Duration
+	httpClient *http.Client
+	apiClient  *client.Client
+	uri        string
+	token      string
+	agentPort  int
+}
+
+func init() {
+	Init()
+}
+
+// Init is exported.
+func Init() {
+	discovery.Register("swarmkit", &Discovery{})
+}
+
+func (s *Discovery) SetAgentPort(discoveryEnginePort int) {
+	s.agentPort = discoveryEnginePort
+}
+
+// Initialize initializes the discovery with the SwarmKit manager's address
+func (s *Discovery) Initialize(uri string, heartbeat time.Duration, ttl time.Duration, _ map[string]string) error {
+	// the uri may be a manager address or a docker socket
+	if uri == "" {
+		return errors.New("Swarm Mode manager address not provided")
+	}
+	s.uri = uri
+	s.heartbeat = heartbeat
+	s.ttl = ttl
+
+	// create a client to connect with the Swarm Mode API
+	httpClient, _, err := cluster.NewHTTPClientTimeout("tcp://"+uri, nil, time.Duration(30*time.Second), nil)
```

Comment:
So with this approach, each Swarm manager is tied to a specific Swarmkit manager, right? And if that Swarmkit manager goes down, discovery will no longer work on this manager, right?